File: macos/Classes/Models/WhisperTokenData.swift
import Foundation

struct WhisperTokenData {
    var id: Int32
    var tid: Int32
    var p: Float
    var plog: Float
    var pt: Float
    var ptsum: Float
    var t0: Int64
    var t1: Int64
    var vlen: Float

    init(token_data: whisper_token_data) {
        self.id = token_data.id
        self.tid = token_data.tid
        self.p = token_data.p
        self.plog = token_data.plog
        self.pt = token_data.pt
        self.ptsum = token_data.ptsum
        self.t0 = token_data.t0
        self.t1 = token_data.t1
        self.vlen = token_data.vlen
    }

    func toDictionary() -> [String: Any] {
        return [
            "id": self.id,
            "tid": self.tid,
            "p": self.p,
            "plog": self.plog,
            "pt": self.pt,
            "ptsum": self.ptsum,
            "t0": self.t0,
            "t1": self.t1,
            "vlen": self.vlen
        ]
    }
}

File: lib/src/models/whisper_result/whisper_token_data.dart
import 'package:whisper_cpp/whisper_cpp.dart';

class WhisperTokenData {
  final int id;
  final int tid;
  final double p;
  final double plog;
  final double pt;
  final double ptsum;
  final int t0;
  final int t1;
  final double vlen;

  WhisperTokenData({
    required this.id,
    required this.tid,
    required this.p,
    required this.plog,
    required this.pt,
    required this.ptsum,
    required this.t0,
    required this.t1,
    required this.vlen,
  });

  factory WhisperTokenData.fromJson(Map<Object?, Object?>? data) {
    Map<String, dynamic> json = Map<String, dynamic>.from(data as Map);

    return WhisperTokenData(
      id: IntHandler.parse(json['id']),
      tid: IntHandler.parse(json['tid']),
      p: DoubleHandler.parse(json['p']),
      plog: DoubleHandler.parse(json['plog']),
      pt: DoubleHandler.parse(json['pt']),
      ptsum: DoubleHandler.parse(json['ptsum']),
      t0: IntHandler.parse(json['t0']),
      t1: IntHandler.parse(json['t1']),
      vlen: DoubleHandler.parse(json['vlen']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'tid': tid,
      'p': p,
      'plog': plog,
      'pt': pt,
      'ptsum': ptsum,
      't0': t0,
      't1': t1,
      'vlen': vlen,
    };
  }
}
